src/build/common/inlineCache.js

```diff
@@ -10,45 +12,62 @@ var exts = {
 };
 
 exports.InlineCache = Class(function () {
-  this.init = function () {
+  this.init = function (logger) {
     this._cache = {};
-  }
-
-  this.addFiles = function (files, cb) {
-    var f = ff(this, function () {
-      for (var i = 0, n = files.length; i < n; ++i) {
-        this.add(files[i], f.wait());
-      }
-
-      f(this);
-    }).cb(cb);
-  }
+    this._logger = logger;
+  };
 
   this.has = function (relativePath) {
     return !!this._cache[relativePath];
-  }
-
-  this.add = function (file, cb) {
-    var ext = path.extname(file.fullPath);
-    if (!exts[ext]) { return cb(); }
-
-    var f = ff(this, function () {
-      fs.readFile(file.fullPath, 'utf-8', f());
-    }, function (contents) {
-      if (ext == '.json') {
-        // validate and remove whitespace:
-        try {
-          contents = JSON.stringify(JSON.parse(contents));
-        } catch (e) {
-          console.error('invalid JSON:', file.fullPath);
+  };
+
+  // returns true if it doesn't add it so that it can be used with a filter
+  // function (filter out the files that get inlined)
+  this.add = function (file) {
+    var ext = path.extname(file.path);
+    if (!exts[ext] || file.inline === false) { return true; }
+
+    return new Promise(function (resolve, reject) {
+        if (file.contents) {
+          resolve(file.contents);
+        } else {
+          readFile(file.history[0], 'utf8').then(resolve, reject);
+        }
+      })
+      .bind(this)
+      .then(function (contents) {
+        if (contents instanceof stream.Stream) {
+          return streamToArray(contents);
+        } else {
+          return contents;
         }
-      }
+      })
+      .then(function (contents) {
+        if (Array.isArray(contents)) {
+          contents = Buffer.concat(contents);
+        }
+
+        if (ext === '.json') {
+          try {
+            contents = JSON.stringify(JSON.parse(contents.toString('utf8')));
+          } catch (e) {
+            this._logger.warn('invalid JSON in', file.path);
+          }
+        }
+
+        this._cache[file.relative] = contents;
 
-      this._cache[file.target] = contents;
-    }).cb(cb);
-  }
+        return false;
+      })
+      .catch(function (e) {
+        // ignore errors inlining, just copy the file
+        this._logger.warn('inline cache failed to process file, falling back to'
+                  + ' copying:', file.path, e.stack);
```

Comment:
I'm not entirely sure under what conditions this would fail, so I'm not 100% sure what the right error message is.  It's possible we can just let the error fall all the way through and stop the build... I think it's an unexpected error?